servers/src/main/java/tachyon/worker/block/BlockLockManager.java

```diff
@@ -0,0 +1,225 @@
+/*
+ * Licensed to the University of California, Berkeley under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package tachyon.worker.block;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.Lock;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Sets;
+
+import tachyon.Constants;
+
+/**
+ * Handle all block locks.
+ * <p>
+ * This class is thread-safe.
+ */
+public class BlockLockManager {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+  /** The number of locks, larger value leads to finer locking granularity, but more space. */
+  // TODO: Make this configurable
+  private static final int NUM_LOCKS = 100;
+  /** The unique id of each lock */
+  private static final AtomicLong LOCK_ID_GEN = new AtomicLong(0);
+
+  private final BlockMetadataManager mMetaManager;
+  /** A map from a block ID to its lock */
+  private final ClientRWLock[] mLockArray = new ClientRWLock[NUM_LOCKS];
+  /** A map from a user ID to all the locks hold by this user */
+  private final Map<Long, Set<Long>> mUserIdToLockIdsMap = new HashMap<Long, Set<Long>>();
+  /** A map from a lock ID to the lock record of it */
+  private final Map<Long, LockRecord> mLockIdToRecordMap = new HashMap<Long, LockRecord>();
+  /** To guard access on mLockIdToRecordMap and mUserIdToLockIdsMap */
+  private final Object mSharedMapsLock = new Object();
+
+  public BlockLockManager(BlockMetadataManager metaManager) {
+    mMetaManager = Preconditions.checkNotNull(metaManager);
+    for (int i = 0; i < NUM_LOCKS; i ++) {
+      mLockArray[i] = new ClientRWLock();
+    }
+  }
+
+  /**
+   * Lock a block if it exists, throw IOException otherwise.
+   *
+   * @param userId the ID of user
+   * @param blockId the ID of block
+   * @param blockLockType READ or WRITE
+   * @return lock id if the block exists
+   * @throws IOException
+   */
+  public long lockBlock(long userId, long blockId, BlockLockType blockLockType) throws IOException {
+    // TODO: generate real hashValue on blockID.
+    int hashValue = (int) (blockId % (long) NUM_LOCKS);
```

Comment:
This should probably be changed, doesn't need to be anything fancy, but I think due to how blockIds are calculated, almost all blocks will be mapped to a few buckets (the only way this would be uniform is if each file always had 100 blocks).